R/install.R

```diff
@@ -0,0 +1,264 @@
+    utils::compareVersion(python_info$version, MIN_PYTHON_VERSION) >= 0) {
+    return(TRUE)
+  }
+  return(FALSE)
+}
+
+python_versions_in_path <- function() {
+  python_regex <- "python(3|3.\\d)?$"
+  if (Sys.info()[["sysname"]] == "Windows") {
+    python_regex <- "python(3|3.\\d)?.exe$"
+  }
+
+  found_versions <- NULL
+  path_options <- strsplit(Sys.getenv("PATH"), split = ":")
+  for (path in path_options) {
+    path_versions <- list.files(
+      path = path,
+      pattern = python_regex,
+      full.names = TRUE
+    )
+    found_versions <- c(found_versions, path_versions)
+  }
+
+  return(found_versions)
+}
+
+find_compatible_python <- function() {
+  # Version to be installed on Windows if no compatible python found.
+  PYTHON_VERSION_TO_BE_INSTALLED <- "3.8"
+
+  python_path <- NULL
+
+  # If RETICULATE_PYTHON is set, check that Python first.
+  preset_python <- Sys.getenv("RETICULATE_PYTHON")
+  if (preset_python != "") {
+    if (python_is_compatible(preset_python)) {
+      Sys.setenv(RETICULATE_PYTHON = preset_python)
+      return(preset_python)
+    }
+    # if python isn't compatible, RETICULATE_PYTHON is unset
+  }
+
+  # List all versions of Python found, from py_discover_config().
+  # If on Mac/Linux, also search along $PATH since reticulate doesn't auto-search.
+  sysname <- Sys.info()[["sysname"]]
+  config_info <- reticulate::py_discover_config()
+  all_versions <- config_info$python_versions
+  if (sysname == "Darwin" || sysname == "Linux") {
+    all_versions <- c(all_versions, python_versions_in_path())
+  }
+
+  for (found_python in all_versions) {
+    if (python_is_compatible(found_python)) {
+      python_path <- found_python
+      break
+    }
+  }
+
+  # If no compatible Python found, try some backup options and/or display a
+  # custom error message depending on OS.
+  if (is.null(python_path)) {
+    error_msg <- paste(
+      "No compatible version of Python (>=3.6) with development libraries ",
+      "(libpython) found, which are needed for parallelization.",
+      "",
+      "Ensure that a compatible version of Python is on your PATH env variable.",
+      "",
+      "If problems persist, set RETICULATE_PYTHON manually and re-install:",
+      "",
+      "  Sys.setenv(RETICULATE_PYTHON = \"/path/to/your/python\")",
+      "  toolchest::install_toolchest()",
+      sep = "\n"
+    )
+    packageStartupMessage(error_msg)
+
+    # if on MacOS, suggest installing python3 via homebrew
+    if (sysname == "Darwin") {
+      suggestion_message <- paste(
+        "MacOS detected. Try installing Python3 (>=3.6) via Homebrew (https://brew.sh/):",
+        "",
+        "  brew install python",
+        "",
+        "If Homebrew + Python are installed and problems persist, try reinstalling:",
+        "",
+        "  brew reinstall python",
+        "",
+        "After installation, make sure that Homebrew is on your PATH env var.",
+        "If problems persist, set RETICULATE_PYTHON manually and re-install:",
```

Comment:
These sound like a scary couple lines for someone whose furthest foray into coding is R.

I see below that you specify how to set RETICULATE_PYTHON manually, but there's a chance that people don't even know what their PATH is or how to check it. Maybe we could link to a GitHub issue where we include these instructions, and say something along the lines of "If you're still having issues, contact us at Toolchest or check out github.com/issue/xxx"?

Reply:
Sure, I'll add some instructions there. (I don't think line about the PATH is strictly necessary anymore.)